src/rabbit_delayed_message.erl

```diff
@@ -260,3 +267,17 @@ recover_exchange_and_bindings(#exchange{name = XName} = X) ->
                               "recovered bindings for ~s",
                              [rabbit_misc:rs(XName)])
     end).
+
+bump_routed_stats(ExName, Qs) ->
+    rabbit_global_counters:messages_routed(amqp091, length(Qs)),
+    [begin
+         QName = amqqueue:get_name(Q),
+         Key = {QName, ExName},
+         rabbit_core_metrics:channel_stats(queue_exchange_stats, publish, {self(), Key}, 1),
+         %% Keys in the process dictionary are used to clean up the core metrics
+         %% FIXME see rabbit_channel:erase_queue_stats/1
+         %%put({queue_exchange_stats, Key}, none)
```

Comment:
should the delayed exchange process monitor all queues it ever published to and mark the metric when any of them is deleted?
Or would it cause problems if these metrics were never deleted?

Reply:
Monitoring every queue will be expensive. Keeping these samples around will leak
resources.

This exchange doesn't have access to the context. Since no other exchange is stateful and tries to do the work channels usually would, there was never any need to propagate the channel to `route/2`.

Reply:
Before I assumed that the metrics GC process uses the delete markers but now that I look at it, to me it looks like it diligently scans the whole table and removes all entries of deleted queues/exchanges it finds, including those that would be added by this PR (https://github.com/rabbitmq/rabbitmq-server/blob/master/deps/rabbit/src/rabbit_core_metrics_gc.erl#L191-L196) In this case there would be no metric leak and no need to monitor all queues. Do I see this correctly? What are metrics delete markers used for? I can't find usage in current master. Maybe a left-over from a previous metrics implementation?

Reply:
Ah, I forgot about that. It was added after this metrics design shipped, as a failsafe. Can you please verify that indeed leftover rows are always removed so this exchange can basically pass any value for an ID?

Also, if we pass a pid of a live process, would those rows be cleared? This is the question.

Reply:
did some tests of what happens when a delayed exchange or queue for which there are metrics is deleted

`rabbit_core_metics`, table: `channel_queue_exchange_metrics`
1. at delayed delivery a row with key "FakeChannelPid"+exchange+queue is written on the node of the exchange (where the original publish happened)
2. when a queue is deleted it calls `rabbit_core_metrics:queue_deleted/1` to mark all rows deleted for that queue on the node of the queue (might or might not be the same as publish node)
3. channels also monitor queues and call `rabbit_core_metrics:channel_queue_exchange_down/1` to mark entries for that queue as deleted (on the channel node) - this does not happen if the delay exchange does the publishing
4. `rabbit_mgmt_metrics_collector:aggregate_entry` will delete rows which are marked as deleted. This won't happen if a queue is deleted on another node than the publish.
5. there is a failsafe GC you mentioned that periodically goes through all entries and deletes (not just marks) those entries for which the queue or exchange does not exist or the Pid is not alive (this works for both channel or delay exchange PID). This makes sure there are no entries left behind even if 3./4. does not do its job. (This happens after a delay, not immediately when the resource is deleted.)

Additionally in case `rates_mode=detailed`
6. `rabbit_mgmt_metrics_collector:aggregate_entry` will call `insert_entry_ops(queue_exchange_stats_publish, ...)` if the core_metrics row is not marked for delete. However for a deleted exchange or queue this metric won't increase and the entries in `queue_exchange_stats_publish` table will expire/rotate out (I think this is done by `exometer_slide`)

> Also, if we pass a pid of a live process, would those rows be cleared? This is the question.
If the delayed exchange and the queue are still defined the row won't be cleared but I think this is expected. The `rabbit_delayed_message` gen_server process acts as a long running channel from this point of view.

So to sum up I couldnt find any leaking ETS row that would not be cleaned up after an exchange or queue is deleted.